x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -281,7 +291,10 @@ private static Version readMappingVersion(String indexName, MappingMetaData mapp
      */
     public void checkIndexVersionThenExecute(final Consumer<Exception> consumer, final Runnable andThen) {
         final State indexState = this.indexState; // use a local copy so all checks execute against the same state!
-        if (indexState.indexExists && indexState.isIndexUpToDate == false) {
+        if (indexState.concreteIndexName == null) {
+            // state not yet recovered
+            delayUntilStateRecovered(consumer, () -> checkIndexVersionThenExecute(consumer, andThen));
```

Comment:
This worries me.
We're suddenly adding in a queue of pending tasks to run when the recovery block is removed, but I don't see any analysis on how big that queue could get, and what the impact is on the thread pool when it happens.

We don't do that now. We don't queue if the security index is red, we currently fail if the index is not recovered.
There's not enough information in this PR to know whether or not it's a sensible thing to do, and how we'd protect ourselves from getting into trouble with it (e.g. too many queued tasks, or delayed read + writes because recovery was slow).

From a usability point of view it's a nice idea, but it feels like we're jumping in without thinking it through.

Reply:
Okay, understood! I just wonder if there is some alternative for "delaying requests" that would be acceptable beyond feature freeze?